adam/learner.py

```diff
@@ -102,3 +106,77 @@ def describe(
+    """
+
+    _descriptions_to_properties: Dict[
+        LinguisticDescriptionT, Set[PropertyPerception]
+    ] = attrib(init=False, default=Factory(dict))
+
+    def observe(
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescriptionT]
+    ) -> None:
+        if len(learning_example.perception.frames) != 1:
+            raise RuntimeError("Subset learner can only handle single frames for now")
+
+        is_new_description = True
+        observed_linguistic_description = learning_example.linguistic_description
+        if isinstance(
+            learning_example.perception.frames[0], DevelopmentalPrimitivePerceptionFrame
+        ):
+            observed_property_assertions = learning_example.perception.frames[
+                0
+            ].property_assertions
+        else:
+            raise RuntimeError("Cannot process perception type.")
+
+        for description in self._descriptions_to_properties.keys():
+            # If already observed, reduce the properties for that description to the common subset of
+            # the new observation and the previous observations
+            if (
+                observed_linguistic_description.as_token_sequence()
+                == description.as_token_sequence()
+            ):
+                self._descriptions_to_properties[
+                    description
+                ] = self._descriptions_to_properties[description].intersection(
+                    observed_property_assertions
+                )
+                is_new_description = False
+        # If it's a new description, learn that as a new observation
+        if is_new_description:
+            self._descriptions_to_properties[observed_linguistic_description] = set(
+                observed_property_assertions
+            )
+
+    def describe(
+        self, perception: PerceptualRepresentation[PerceptionT]
+    ) -> Mapping[LinguisticDescriptionT, float]:
+        if len(perception.frames) != 1:
+            raise RuntimeError("Subset learner can only handle single frames for now")
+        if isinstance(perception.frames[0], DevelopmentalPrimitivePerceptionFrame):
+            observed_property_assertions = perception.frames[0].property_assertions
+        else:
+            raise RuntimeError("Cannot process perception type.")
+
+        # get the learned description for which there are the maximum number of matching properties (i.e. most specific)
+        max_matching_properties = 0
+        learned_description = None
+        for description, properties in self._descriptions_to_properties.items():
+            if all(prop in observed_property_assertions for prop in properties) and (
+                len(properties) > max_matching_properties
+            ):
+                learned_description = description
+                max_matching_properties = len(properties)
+        if learned_description and max_matching_properties > 0:
```

Comment:
If `learned_description` is non-None than `max_matching_properties` is already known to be >0